transport/grpc.go

```diff
@@ -0,0 +1,188 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import (
+	"bytes"
+	"errors"
+	"io/ioutil"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/multierr"
+	apipeer "go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/peer"
+	"go.uber.org/yarpc/peer/hostport"
+	"go.uber.org/yarpc/peer/roundrobin"
+	"go.uber.org/yarpc/transport/grpc"
+	"golang.org/x/net/context"
+)
+
+var (
+	errGRPCNoAddresses = errors.New("must specify at least one grpc address")
+	errGRPCNoTracer    = errors.New("must specify grpc tracer")
+	errGRPCNoCaller    = errors.New("must specify grpc caller")
+	errGRPCNoService   = errors.New("must specify grpc service")
+	errGRPCNoEncoding  = errors.New("must specify grpc encoding")
+	errGRPCNoProcedure = errors.New("must specify grpc procedure")
+)
+
+// GRPCOptions are used to create a GRPC transport.
+type GRPCOptions struct {
+	Addresses       []string
+	Tracer          opentracing.Tracer
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+}
+
+// NewGRPC returns a transport that calls a GRPC service.
+func NewGRPC(options GRPCOptions) (TransportCloser, error) {
+	return newGRPC(options)
+}
+
+type grpcTransport struct {
+	Transport       transport.Transport
+	Outbound        transport.UnaryOutbound
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+	tracer          opentracing.Tracer
+}
+
+func newGRPC(options GRPCOptions) (*grpcTransport, error) {
+	if len(options.Addresses) == 0 {
+		return nil, errGRPCNoAddresses
+	}
+	if options.Tracer == nil {
+		return nil, errGRPCNoTracer
+	}
+	if options.Caller == "" {
+		return nil, errGRPCNoCaller
+	}
+	if options.Encoding == "" {
+		return nil, errGRPCNoEncoding
+	}
+
+	identifiers := make([]apipeer.Identifier, len(options.Addresses))
+	for i, address := range options.Addresses {
+		identifiers[i] = hostport.Identify(address)
+	}
+	transport := grpc.NewTransport(grpc.Tracer(options.Tracer))
+	outbound := transport.NewOutbound(peer.Bind(roundrobin.New(transport), peer.BindPeers(identifiers)))
+	if err := transport.Start(); err != nil {
+		return nil, err
+	}
+	if err := outbound.Start(); err != nil {
+		_ = transport.Stop()
+		return nil, err
```

Comment:
multierr.Combine?

Reply:
I thought this was fine as adding the error from `Stop` when the `Start` error is what was the real problem could be confusing.

Reply:
+1, although we tend to just do `transport.Stop()`, don't think we use `errcheck` here.